service/pkg/server/services.go

```diff
@@ -19,162 +22,185 @@ import (
-			slog.Bool("configured runMigrations", cfg.DB.RunMigrations),
-			slog.String("reason", reason),
-		)
+// newServiceDBClient creates a new database client for the specified namespace.
+// It initializes the client with the provided context, logger configuration, database configuration,
+// namespace, and migrations. It returns the created client and any error encountered during creation.
+func newServiceDBClient(ctx context.Context, logCfg logger.Config, dbCfg db.Config, ns string, migrations *embed.FS) (*db.Client, error) {
+	var err error
+
+	client, err := db.New(ctx, dbCfg, logCfg,
+		db.WithService(ns),
+		db.WithMigrations(migrations),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("issue creating database client for %s: %w", ns, err)
 	}
 
-	// Create the service
-	impl, handler := s.RegisterFunc(serviceregistry.RegistrationParams{
-		Config:                 cfg.Services[s.Namespace],
-		OTDF:                   otdf,
-		DBClient:               d,
-		SDK:                    client,
-		WellKnownConfig:        wellknown.RegisterConfiguration,
-		RegisterReadinessCheck: health.RegisterReadinessCheck,
-		Logger:                 logger.With("namespace", s.Namespace),
-	})
-
-	// Register the service with the gRPC server
-	otdf.GRPCServer.RegisterService(s.ServiceDesc, impl)
-
-	// Register the service with in process gRPC server
-	otdf.GRPCInProcess.GetGrpcServer().RegisterService(s.ServiceDesc, impl)
-
-	// Register the service with the gRPC gateway
-	if err := handler(ctx, otdf.Mux, impl); err != nil {
-		logger.Error("failed to start service", slog.String("namespace", s.Namespace), slog.String("error", err.Error()))
-		return s, d, err
-	}
+	return client, nil
+}
 
-	logger.Info("started service", slog.String("namespace", s.Namespace), slog.String("service", s.ServiceDesc.ServiceName))
-	s.Started = true
-	s.Close = func() {
-		if d != nil {
-			logger.Info("closing database client", slog.String("namespace", s.Namespace), slog.String("service", s.ServiceDesc.ServiceName))
-			// TODO: this might be a problem if we can't call close on the db client multiple times
-			d.Close()
-		}
+// determineStatusOfMigration determines the status of the migration based on the provided client.
+// It checks if the client is required, if the required migrations have already been ran,
+// if the service does not require a database, or if the migrations are disabled.
+// It returns a string indicating the reason for the determined status.
+func determineStatusOfMigration(client *db.Client) string {
+	required := (client != nil)
+	requiredAlreadyRan := required && client.MigrationsEnabled() && client.RanMigrations()
+	noDBRequired := !required
+	migrationsDisabled := (required && !client.MigrationsEnabled())
+
+	reason := "undetermined"
+	switch {
+	case requiredAlreadyRan:
+		reason = "required migrations already ran"
+	case noDBRequired:
+		reason = "service does not require a database"
+	case migrationsDisabled:
+		reason = "migrations are disabled"
```

Comment:
Should we capture a failed migration state?

Reply:
This is what we had existing. I just moved stuff around.